vllm/model_executor/models/whisper.py

```diff
@@ -0,0 +1,445 @@
+            config.max_target_positions, self.d_model)
+        self.layers = nn.ModuleList([
+            WhisperDecoderBlock(config, linear_method=linear_method)
+            for _ in range(config.decoder_layers)
+        ])
+        self.layer_norm = nn.LayerNorm(config.d_model)
+
+    def forward(
+        self,
+        input_ids: torch.Tensor,
+        encoder_hidden_states: torch.Tensor,
+        kv_cache: List[KVCache],
+        input_metadata: InputMetadata,
+    ) -> torch.Tensor:
+
+        inputs_embeds = self.embed_tokens(input_ids)
+        positions = self.embed_positions(
+            inputs_embeds,
+            past_key_values_length=0,
+        )
+        hidden_states = inputs_embeds + positions
+        for i, dec_block in enumerate(self.layers):
+            hidden_states = dec_block(hidden_states, encoder_hidden_states,
+                                      kv_cache[i], input_metadata)
+
+        hidden_states = self.layer_norm(hidden_states)
+        return hidden_states
+
+
+class WhisperForConditionalGeneration(nn.Module):
+
+    def __init__(
+        self,
+        config: WhisperConfig,
+        audio_features_config: AudioFeaturesConfig,
+        linear_method: Optional[LinearMethodBase] = None  # probably not needed
+    ):
+        super().__init__()
+        self.config = config
+        self.encoder = WhisperEncoder(config, linear_method)
+        self.decoder = WhisperDecoder(config, linear_method)
+        self.sampler = Sampler(config.vocab_size)
+
+    def forward(
+        self,
+        input_features: torch.FloatTensor,
+        positions: torch.Tensor,
+        kv_caches: List[KVCache],
+        input_metadata: InputMetadata,
+        input_ids: Optional[torch.Tensor] = None,
+    ) -> torch.Tensor:
+
+        input_features = input_features.to(dtype=torch.float32)
+        if input_metadata.is_prompt:
+            # prompt run, need to run encoder once
+            hidden_states = self.encoder(input_features, input_metadata=input_metadata)
+            input_metadata.attn_bias = None
+            
+            if input_ids is None:
+                decoder_input_ids = torch.tensor([[1]]).to(
+                    input_features.device) * self.config.decoder_start_token_id
+            else:
+                decoder_input_ids = input_ids
+
+        if kv_caches[0][0] is None:
+            hidden_states = None
+        else:
+            hidden_states = self.decoder(input_ids=decoder_input_ids,
+                                          encoder_hidden_states=hidden_states,
+                                          kv_cache=kv_caches,
+                                          input_metadata=input_metadata)
+        return hidden_states
+
+    def sample(self, hidden_states: torch.Tensor,
+               sampling_metadata: SamplingMetadata):
+        # TODO: For now we are not implementing the sampling method
+        return hidden_states
+
+    def load_weights(
```

Comment:
Can you match the style of llama in this function? They have been very consistent with this logic across functions

https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py#L363

Reply:
Looking good, seems like we could adhere to their convention